cogeo_mosaic/backends/base.py

```diff
@@ -62,5 +67,83 @@ def write(self):
         """Upload new MosaicJSON to backend."""
 
     @abc.abstractmethod
-    def update(self):
+    def update(
+        self,
+        features: Sequence[Dict],
+        accessor: Callable = DEFAULT_ACCESSOR,
+        overwrite: bool = False,
+        **kwargs: Any,
+    ):
+        """Update existing MosaicJSON on backend."""
+
+    def _update_quadkey(self, quadkey: str, dataset: List[str]):
+        """Update quadkey list."""
+        self.mosaic_def.tiles[quadkey] = dataset
+
+    def _update_metadata(self, bounds: List[float], version: str):
+        """Update bounds and center."""
+        self.mosaic_def.version = version
+        self.mosaic_def.bounds = bounds
+        self.mosaic_def.center = (
+            (bounds[0] + bounds[2]) / 2,
+            (bounds[1] + bounds[3]) / 2,
+            self.mosaic_def.minzoom,
+        )
+
+    def _update(
+        self,
+        features: Sequence[Dict],
+        accessor: Callable,
+        add_first: bool = True,
+        **kwargs,
+    ):
         """Update existing MosaicJSON on backend."""
+        version = list(map(int, self.mosaic_def.version.split(".")))
+        version[-1] += 1
+        version = ".".join(map(str, version))
+
+        updated_quadkeys = set()
+
+        dataset_geoms = polygons(
+            [feat["geometry"]["coordinates"][0] for feat in features]
+        )
+
+        tiles = burntiles.burn(features, self.quadkey_zoom)
+        tiles = [mercantile.Tile(*tile) for tile in tiles]
+
+        tree = STRtree(dataset_geoms)
+
+        for tile in tiles:
+            quadkey = mercantile.quadkey(tile)
+            tile_geom = polygons(mercantile.feature(tile)["geometry"]["coordinates"][0])
+
+            # Find intersections from rtree
+            intersections_idx = sorted(tree.query(tile_geom, predicate="intersects"))
+            if len(intersections_idx) == 0:
+                continue
+
+            intersect_dataset, intersect_geoms = zip(
+                *[(features[idx], dataset_geoms[idx]) for idx in intersections_idx]
+            )
+
+            dataset = self.mosaic_def._filter(
+                tile, intersect_dataset, intersect_geoms, **kwargs
+            )
+
+            new_assets = [accessor(f) for f in dataset]
+
+            assets = self.tile(*tile)
+            assets = [*new_assets, *assets] if add_first else [*assets, *new_assets]
+            self._update_quadkey(quadkey, assets)
+
+            updated_quadkeys.add(tile)
```

Comment:
@kylebarron In fact I just had to copy code from `MosaicJSON._create_mosaic`. This should reduce the number of read/write

Reply:
I was originally going to say that you could probably just let `_create_mosaic` take an optional argument, something like `existing_mosaic`, that defaults to `None`, but if it's a `dict`, you fill the new tiles into that.

That way you aren't duplicating this code, but I'm not sure it's worth it to reduce this one duplication.

Reply:
Well I think this is doable. We can create a mosaic with the new dataset and then merge the `tiles` dict together.

I'll do this when #66 get merged

Reply:
```python
    def update(
        self,
        features: Sequence[Dict],
        add_first: bool = True,
        **kwargs,
    ):
        """Update existing MosaicJSON on backend."""
        version = list(map(int, self.mosaic_def.version.split(".")))
        version[-1] += 1
        version = ".".join(map(str, version))

        new_mosaic = MosaicJSON.from_features(
            features,
            self.mosaic_def.minzoom,
            self.mosaic_def.maxzoom,
            quadkey_zoom=self.quadkey_zoom,
            **kwargs,
        )
        for quadkey, new_assets in new_mosaic.tiles.items():
            tile = mercantile.quadkey_to_tile(quadkey)
            assets = self.tile(*tile)
            assets = [*new_assets, *assets] if add_first else [*assets, *new_assets]

            # add custom sorting algorithm (e.g based on path name)
            self._update_quadkey(quadkey, assets)

        nxmin, nymin, nxmax, nymax = new_mosaic.bounds
        oxmin, oymin, oxmax, oymax = self.mosaic_def.bounds

        bounds = [
            min(nxmin, oxmin), min(nymin, oymin), max(nxmax, oxmax), max(nymax, oymax)
        ]
        self._update_metadata(bounds, version)

        # We only write is path is set
        if self.path:
            self.write()

        return
```